### DISCUSSION TITLE
MeshRelation and barycentric

### DISCUSSION THREAD

**Discussion by stephomi** [elalish/manifold#477] [Q&A]
For now I rely on faceID by iterating on the triangles range through the `runOriginalID` thing, then compute the barycentric myself.
It's working fine.

Context: https://github.com/elalish/manifold/wiki/Manifold-Library#vertex-properties
^ Is this still a thing? I can't seem to access barycentric.

Vertex properties isn't the best fit for me as:
- I can have many properties, and most of them are 8bit, 16bit or packed attributes
- I can have X layer (morphtarget/blendshape) for each objects, and each layer is at least 7 attribute (offset, color, rough, metal and opacity for paint channels), all of them packed or compressed
- And all of that is packed or compressed, unflattening everything in float can exceed RAM limit on mobile.
- In the future I'll probably add rigging which might require custom interpolation (integer bone index logic linked to bone weight).

**Comment by elalish** [comment]
Ah, good call, I need to update the wiki there. Indeed, I removed the barycentric stuff as it turned out not to be working anyway and replaced it with our properties API. If you don't want Manifold handling your properties for you (due to memory and such) you can just assign 3 property channels to act as barycentric coordinates (`[1,0,0]`, `[0,1,0]`, `[0,0,1]` on the verts of each tri) and then remap them afterwards. 

I've never heard of having so many vertex properties on a mesh before - I figured generally you'd just have a UV map and do the rest with textures.

**Reply by elalish** [reply]
I've updated the wiki: https://github.com/elalish/manifold/wiki/Manifold-Library#vertex-properties

**Comment by stephomi** [comment]
Thought about that but the un-indexing means I basically multiply vertex count by 4-6 (depending on the tri/quad ratio) and it was not necessarily easier than recomputing the barycentric. **Anyway, it's not an issue**, just wanted to make sure I was understanding the doc correctly.
I rely heavily on faceID to reconstruct quad, and it's working great.

For now I drop the UVs, but I'd like to keep them, I suppose that's what the `mergeFromVert/mergeToVert` is for.

---
As for vertex properties, the context is [digital sculpting](https://apps.apple.com/us/app/id1519508653?mt=8&platform=ipad), where most of the time you don't have UV since the topology changes regularly (voxel remeshing, dynamic tesselation, trim, etc).
But material painting put aside, morph target are pretty common in 3d.
But softwares easily drop those data if it's a pain to keep it  (same with UVs).

**Reply by elalish** [reply]
Great, glad it's working for you! And you're exactly right regarding UVs. Thanks for the context, that's very cool. Are you making use of our `Warp` function by chance? Also, please feel free to share anything you build that uses Manifold in #340 - open and closed are equally welcome.

**Reply by stephomi** [reply]
I don't use Warp, I'm not sure in which scenario it should be used?


---
For information, to reconstruct the quads I notice manifold doesn't keep the order (v1,v2,v3).
So I use this snippet to re-order it when matching the other side:
```cpp
static glm::ivec4 makeQuad(glm::ivec4 &a, glm::ivec4 &b) {
    if (a.x == b.x) {
        if (a.y == b.z) return glm::ivec4(b.y, a.y, a.z, b.x);
        if (a.z == b.y) return glm::ivec4(a.x, a.y, a.z, b.z);
    } else if (a.x == b.y) {
        if (a.y == b.x) return glm::ivec4(b.z, a.y, a.z, b.y);
        if (a.z == b.z) return glm::ivec4(a.x, a.y, a.z, b.x);
    } else if (a.x == b.z) {
        if (a.y == b.y) return glm::ivec4(b.x, a.y, a.z, b.z);
        if (a.z == b.x) return glm::ivec4(a.x, a.y, a.z, b.y);
    } else if (a.y == b.x) {
        if (a.z == b.z) return glm::ivec4(a.x, a.y, b.y, b.z);
    } else if (a.y == b.y) {
        if (a.z == b.x) return glm::ivec4(a.x, a.y, b.z, b.x);
    } else if (a.y == b.z) {
        if (a.z == b.y) return glm::ivec4(a.x, a.y, b.x, b.y);
    }
    return glm::ivec4(-1);
}
```

I'm still working on the quad thing around the intersection area (testing whether or not I need to check consistent orientation or making concavity check, etc).
Example with a concave check (top row).
<img width="587" alt="Capture d’écran 2023-07-03 à 19 29 00" src="https://github.com/elalish/manifold/assets/4750281/0aa32320-53f5-4336-a7c0-a94033a90a31">

**Reply by elalish** [reply]
You mentioned sculpting - `Warp` is a very general concept so it can be used for lots of different things that don't change the topology. It's great for sweeps, but your imagination is the only limit. It'll be better when I get a self-intersection removal algorithm in.

**Comment by stephomi** [comment]
[Gist](https://gist.github.com/stephomi/2c3476e1322790cad6b1da518677d787) on how to reconstruct quads.
I got asked how I did it, so I figured it might interest others.

Not sure if that's the simplest way.
